src/azul/auth.py

```diff
@@ -12,6 +24,42 @@ def identity(self) -> str:
         """
         raise NotImplementedError
 
+    _cls_field = '_cls'
+
+    def to_json(self) -> JSON:
+        """
+        >>> @attr.s(auto_attribs=True)
+        ... class Foo(Authentication):
+        ...     foo: str
+        ...     def identity(self) -> str:
+        ...         # noinspection PyUnresolvedReferences
+        ...         return self.foo
+        >>> f = Foo('bar')
+        >>> f
+        Foo(foo='bar')
+        >>> f.to_json()
+        {'foo': 'bar', '_cls': 'Foo'}
+        >>> Authentication.from_json(f.to_json())
+        Foo(foo='bar')
+        """
+        assert self._cls_field not in dir(self), self
```

Comment:
`asdict` will only translate declared fields so this is overly restrictive, and expensive. Consider checking this statically, and using `attrs.fields`.